### PR TITLE
WIP: split `.js`, load some asynchronously & on-demand, use ESM

### PR DESCRIPTION
**Pull request by WofWca**
## Proposed changes

This is super WIP, for now look at it as a discussion rather than a full-fledged MR.

In order to improve page loading performance, it would be cool to load some JS asynchronously and on-demand, and tree-shake some of it. Examples: shortcuts, syntax highlighting, username autocompletion.
Here's the audit I ran:

image

As you can see, on my PC it takes ~400ms to load a __prefetched__ page (i.e. it's only rendering, not including network download). You can imagine it's a little worse on mobile.

First step to this is rewriting the scripts with imports.

This also makes the code easier to maintain as there's fewer global vars.

TODO:
* [ ] Make sure nothing broke
* [ ] See TODOs in the code
* [ ] Perhaps a transpiler needs to be used as [ESM imports are not supported in old browsers (~92% now)](https://caniuse.com/es6-module-dynamic-import). Which one? Webpack?

## Checklist

- [ ] Lint and unit tests pass locally with my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added documentation to describe my feature.
- [ ] I have squashed my commits into logic units.
- [ ] I have described the changes in the commit messages.

### CODE REVIEW

File: weblate/templates/base.html [hunk 1]
```diff
@@ -62,23 +63,9 @@
...
 </script>
 {% endif %}
-  <script defer data-cfasync="false" src="{% static 'vendor/jquery.js' %}{{ cache_param }}"></script>
-  <script defer data-cfasync="false" src="{% static 'vendor/js.cookie.js' %}{{ cache_param }}"></script>
-  <script defer data-cfasync="false" src="{% static 'vendor/autosize.js' %}{{ cache_param }}"></script>
-  <script defer data-cfasync="false" src="{% static 'vendor/multi.js' %}{{ cache_param }}"></script>
-  <script defer data-cfasync="false" src="{% static 'vendor/mousetrap.js' %}{{ cache_param }}"></script>
-  <script defer data-cfasync="false" src="{% static 'vendor/mousetrap-global-bind.js' %}{{ cache_param }}"></script>
-  <script defer data-cfasync="false" src="{% static 'vendor/clipboard.js' %}{{ cache_param }}"></script>
+  <!-- TODO separate bootstrap? https://getbootstrap.com/docs/5.1/getting-started/javascript/#using-bootstrap-as-a-module -->
```
Comment (nijel) [inline weblate/templates/base.html:66]: We're still at Bootstrap 3.4.1, see https://github.com/WeblateOrg/weblate/issues/2916

File: weblate/static/editor/full.js [hunk 2]
```diff
@@ -1,3 +1,10 @@
+import $ from "../vendor/jquery.js";
+import Cookies from "./vendor/js.cookie.js";
+import autosize from "./vendor/autosize.js";
+// TODO perf: import Mousetrap dynamically.
+import Mousetrap from "./vendor/mousetrap.js";
```
Comment (nijel) [inline weblate/static/editor/full.js:5]: Mousetrap is IMHO used on nearly every page - it is used to catch Ctrl+Enter for form submission.

**Review by nijel — COMMENTED** [review]
Generally, any improvements to the JavaScript are welcome. The tricky part is that there are basically no tests for it, so it's super easy to break something. The current split out editor files are a typical example of that - it was tested by the author, me when merging and several issues were discovered months after merging in some not that frequently used features.

I think imports is something we can go for directly without a compatibility layer - we already started to use [arrow functions](https://caniuse.com/arrow-functions) two years ago and that was IMHO bigger drop in supported browsers than this.

File: weblate/static/editor/full.js (hunk 2 above)
Comment (WofWca) [reply weblate/static/editor/full.js:5]: Yeah, I saw that. Still, adding a bind asynchronously is possible.

**Comment by WofWca** [comment]
A problem here is that the dependencies don't use ESM (native `export` and `import`), they do UMD or something else, so if you attempt to import them natively, the browser says that "the imported module doesn't provide export named ...".
So looks like we can't do without a transpiler.

**Pull request closed without merging**